### PR TITLE
font-feature-settings gecko glue code

### LINKED ISSUE #15975: Add gecko glue for font-feature-settings property
Servo has had parsing and serialization code for `font-feature-settings` property, but we need to add glue to make it usable in Gecko.

The existing parsing and serialization code is https://github.com/servo/servo/blob/d30978246162e8a9bfad42b43ab6a05c90421c6e/components/style/properties/longhand/font.mako.rs#L634-L742

The glue code should be added under https://github.com/servo/servo/blob/d30978246162e8a9bfad42b43ab6a05c90421c6e/components/style/properties/gecko.mako.rs#L1165-L1167

It seems to me that the existing property code in font.mako.rs also needs some tweak. The feature tag can be represented as a `u32`, so we should not use `String` for it. Also in Gecko's style struct, `gfxFontFeature::mTag` is a `u32` as well.

The target struct can be find in components/style/gecko_bindings/struct_debug.rs (GitHub doesn't display this file because it's too large). It should be `nsFont::fontFeatureSettings`. It is an `nsTArray`, but `gfxFontFeature` seems to be a POD type, so we can use `nsTArray::set_len_pod` for initializing its length.

### PR DESCRIPTION
**Pull request by ghost**
FeatureTagValue value property changed to use u32. ToCss for
FeatureTagValue changed to allow conversion from u32 to string. Parse
for the same struct updated to convert from string to u32. Added two
functions to transfer settings to gecko and copy settings.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #15975  (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16568)

### CODE REVIEW

**Comment by WholeGrainGoats** [comment]
My mistake. I'll see those two errors fixed now.

**Comment by WholeGrainGoats** [comment]
Need to update tests. Need to sleep first, though. Will update the tests to use u32 instead of string to reflect update.

**Comment by Wafflespeanut** [comment]
No need to close the PR for that. Feel free to ping me when it's ready :)

File: components/style/properties/longhand/font.mako.rs [hunk 1]
```diff
@@ -1813,22 +1820,31 @@ ${helpers.single_keyword_system("font-variant-position",
                     return Err(())
                 }
 
+                let b_tag = tag.deref().as_bytes();
```
Comment (Wafflespeanut) [inline components/style/properties/longhand/font.mako.rs]: Doesn't it auto-deref when you directly use `as_bytes`?

File: components/style/properties/longhand/font.mako.rs [hunk 2]
```diff
@@ -1813,22 +1820,31 @@ ${helpers.single_keyword_system("font-variant-position",
...
+                let mut buff: [u8;4] = [0;4];
+
+                for i in 0..4 {
```
Comment (Wafflespeanut) [inline components/style/properties/longhand/font.mako.rs]: You could just zip here like `buff.iter_mut().zip(b_tag.as_bytes())` which will return once the lengths mismatch.

**Review by ghost — CHANGES_REQUESTED** [review]
Thanks for doing this! I have some comments :)

File: components/style/properties/longhand/font.mako.rs (hunk 1 above)
Comment (WholeGrainGoats) [inline components/style/properties/longhand/font.mako.rs]: It had a type error when I attempted to compile without deref. Stated that COW did not have as_bytes defined on it. Adding an explicit deref made the error go away.

**Comment by WholeGrainGoats** [comment]
Will try to fix these tomorrow. Sorry for taking so much time. I should have more free time after finals this Thursday.

File: components/style/properties/longhand/font.mako.rs (hunk 1 above)
Comment (WholeGrainGoats) [inline components/style/properties/longhand/font.mako.rs]: I'll double check, though. Just to be sure.

**Comment by Wafflespeanut** [comment]
No worries! No rush! :)

File: components/style/properties/longhand/font.mako.rs [hunk 3]
```diff
@@ -1792,10 +1792,15 @@ ${helpers.single_keyword_system("font-variant-position",
 
         impl ToCss for FeatureTagValue {
             fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+                use std::{mem,str};
+
+                let raw: [u8;4] = unsafe{mem::transmute(self.tag)};
```
Comment (Manishearth) [inline components/style/properties/longhand/font.mako.rs]: Please don't use unsafe code for stuff like this. Either reconstruct it from the mask, or use the byteorder crate to get a safe version of this transmute.

**Comment by Wafflespeanut** [comment]
@WholeGrainGoats Hey, did you have the time look into this?

**Comment by WholeGrainGoats** [comment]
Looking at it today. Thanks for all the suggestions!

**Comment by WholeGrainGoats** [comment]
@Wafflespeanut Added the byteorder crate to the style tests. Should I remove it and write the u32 equivalent of the string conversion in the file? I added it originally to make sure that we could run this test on big and little endian architectures. As it stands the checks are failing because of the inclusion of this crate in the test's toml file.

**Comment by WholeGrainGoats** [comment]
@Wafflespeanut Added the byteorder crate to the style tests. Should I remove it and write the u32 equivalent of the string conversion in the file? I added it originally to make sure that we could run this test on big and little endian architectures. As it stands the checks are failing because of the inclusion of this crate in the test's toml file.

**Comment by Manishearth** [comment]
You forgot to commit the lockfile change. Apply this diff:

```diff
diff --git a/Cargo.lock b/Cargo.lock
index 9e00ee76e..e01e74c12 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -2799,6 +2799,7 @@ name = "style_tests"
 version = "0.0.1"
 dependencies = [
  "app_units 0.4.0 (registry+https://github.com/rust-lang/crates.io-index)",
+ "byteorder 1.0.0 (registry+https://github.com/rust-lang/crates.io-index)",
  "cssparser 0.13.1 (registry+https://github.com/rust-lang/crates.io-index)",
  "euclid 0.11.3 (registry+https://github.com/rust-lang/crates.io-index)",
  "html5ever 0.16.0 (registry+https://github.com/rust-lang/crates.io-index)",
```

File: components/style/properties/gecko.mako.rs [hunk 4]
```diff
@@ -1302,6 +1302,44 @@ fn static_assert() {
...
+
+        let current_settings = &mut self.gecko.mFont.fontFeatureSettings;
+        let feature_settings = other.gecko.mFont.fontFeatureSettings;
+        let settings_length = future_settings.len();
+
+        current_settings.clear_pod();
+        current_settings.set_len_pod(settings_length);
+
+        for (current, feature) in current_settings.iter_mut().zip(feature_settings) {
+            *current.mTag = feature.tag;
```
Comment (Manishearth) [inline components/style/properties/gecko.mako.rs]: How does this compile? Shouldn't it be `mTag` and `mValue` on both sides?

Comment (Manishearth) [inline components/style/properties/gecko.mako.rs]: Your code has to compile with `./mach build-geckolib` as well as `./mach test-unit -p style`

**Comment by Manishearth** [comment]
This diff should fix your compile issues

```diff
diff --git a/components/style/properties/gecko.mako.rs b/components/style/properties/gecko.mako.rs
index 9a69ba8..c7bf3f1 100644
--- a/components/style/properties/gecko.mako.rs
+++ b/components/style/properties/gecko.mako.rs
@@ -1296,28 +1296,29 @@ fn static_assert() {
     skip_font_longhands = """font-family font-size font-size-adjust font-weight
                              font-synthesis -x-lang font-variant-alternates
                              font-variant-east-asian font-variant-ligatures
-                             font-variant-numeric font-language-override"""
+                             font-variant-numeric font-language-override
+                             font-feature-settings"""
 %>
 <%self:impl_trait style_struct_name="Font"
     skip_longhands="${skip_font_longhands}"
     skip_additionals="*">
 
-    pub fn set_font_feature_settings(&mut self, v: longhands::font_feature_settings::T) {
-        use properties::longhands::font_feature_settings::computed_value::FeatureTagValue;
+    pub fn set_font_feature_settings(&mut self, v: longhands::font_feature_settings::computed_value::T) {
+        use properties::longhands::font_feature_settings::computed_value::{FeatureTagValue, T};
         use gecko_bindings::structs::{gfxFontVariation, nsTArray, gfxFontFeature};
 
         let current_settings = &mut self.gecko.mFont.fontFeatureSettings;
         current_settings.clear_pod();
 
         match v {
-            Normal => current_settings.set_len_pod(0),
+            T::Normal => current_settings.set_len_pod(0),
 
-            Tag(feature_settings) => {
+            T::Tag(feature_settings) => {
                 current_settings.set_len_pod(feature_settings.len());
 
                 for (current, feature) in current_settings.iter_mut().zip(feature_settings) {
-                    *current.mTag = feature.tag;
-                    *current.mValue = feature.value;
+                    current.mTag = feature.tag;
+                    current.mValue = feature.value;
                 }
             }
         };
@@ -1329,14 +1330,14 @@ fn static_assert() {
 
         let current_settings = &mut self.gecko.mFont.fontFeatureSettings;
         let feature_settings = other.gecko.mFont.fontFeatureSettings;
-        let settings_length = future_settings.len();
+        let settings_length = feature_settings.len();
 
         current_settings.clear_pod();
         current_settings.set_len_pod(settings_length);
 
         for (current, feature) in current_settings.iter_mut().zip(feature_settings) {
-            *current.mTag = feature.tag;
-            *current.mValue = feature.value;
+            *current.mTag = feature.mTag;
+            *current.mValue = feature.mValue;
         }
     }
 
diff --git a/components/style/properties/longhand/font.mako.rs b/components/style/properties/longhand/font.mako.rs
index c29ed52..761dd95 100644
--- a/components/style/properties/longhand/font.mako.rs
+++ b/components/style/properties/longhand/font.mako.rs
@@ -1754,7 +1754,7 @@ ${helpers.single_keyword_system("font-variant-position",
                                 spec="https://drafts.csswg.org/css-fonts/#propdef-font-variant-position",
                                 animation_value_type="none")}
 
-<%helpers:longhand name="font-feature-settings" products="none" animation_value_type="none" extra_prefixes="moz"
+<%helpers:longhand name="font-feature-settings" products="gecko" animation_value_type="none" extra_prefixes="moz"
                    spec="https://drafts.csswg.org/css-fonts/#propdef-font-feature-settings">
     use std::fmt;
     use style_traits::ToCss;
```

**Comment by Manishearth** [comment]
Actually, it only fixes some of them, hold on

**Comment by Manishearth** [comment]
This patch should make it compile


```patch
diff --git a/components/style/properties/gecko.mako.rs b/components/style/properties/gecko.mako.rs
index 9a69ba8..b06af8c 100644
--- a/components/style/properties/gecko.mako.rs
+++ b/components/style/properties/gecko.mako.rs
@@ -1296,47 +1296,45 @@ fn static_assert() {
     skip_font_longhands = """font-family font-size font-size-adjust font-weight
                              font-synthesis -x-lang font-variant-alternates
                              font-variant-east-asian font-variant-ligatures
-                             font-variant-numeric font-language-override"""
+                             font-variant-numeric font-language-override
+                             font-feature-settings"""
 %>
 <%self:impl_trait style_struct_name="Font"
     skip_longhands="${skip_font_longhands}"
     skip_additionals="*">
 
-    pub fn set_font_feature_settings(&mut self, v: longhands::font_feature_settings::T) {
-        use properties::longhands::font_feature_settings::computed_value::FeatureTagValue;
-        use gecko_bindings::structs::{gfxFontVariation, nsTArray, gfxFontFeature};
+    pub fn set_font_feature_settings(&mut self, v: longhands::font_feature_settings::computed_value::T) {
+        use properties::longhands::font_feature_settings::computed_value::T;
 
         let current_settings = &mut self.gecko.mFont.fontFeatureSettings;
         current_settings.clear_pod();
 
         match v {
-            Normal => current_settings.set_len_pod(0),
+            T::Normal => unsafe { current_settings.set_len_pod(0) },
 
-            Tag(feature_settings) => {
-                current_settings.set_len_pod(feature_settings.len());
+            T::Tag(feature_settings) => {
+                unsafe { current_settings.set_len_pod(feature_settings.len() as u32) };
 
                 for (current, feature) in current_settings.iter_mut().zip(feature_settings) {
-                    *current.mTag = feature.tag;
-                    *current.mValue = feature.value;
+                    current.mTag = feature.tag;
+                    current.mValue = feature.value;
                 }
             }
         };
     }
 
-    pub fn copy_font_feature_settings_from(&mut self, other: &Self ) {
-        use properties::longhands::font_feature_settings::computed_value::FeatureTagValue;
-        use gecko_bindings::structs::{gfxFontVariation, nsTArray, gfxFontFeature};
+    pub fn copy_font_feature_settings_from(&mut self, other: &Self) {
 
         let current_settings = &mut self.gecko.mFont.fontFeatureSettings;
-        let feature_settings = other.gecko.mFont.fontFeatureSettings;
-        let settings_length = future_settings.len();
+        let ref feature_settings = other.gecko.mFont.fontFeatureSettings;
+        let settings_length = feature_settings.len();
 
         current_settings.clear_pod();
-        current_settings.set_len_pod(settings_length);
+        unsafe { current_settings.set_len_pod(settings_length as u32) };
 
-        for (current, feature) in current_settings.iter_mut().zip(feature_settings) {
-            *current.mTag = feature.tag;
-            *current.mValue = feature.value;
+        for (current, feature) in current_settings.iter_mut().zip(feature_settings.iter()) {
+            current.mTag = feature.mTag;
+            current.mValue = feature.mValue;
         }
     }
 
diff --git a/components/style/properties/longhand/font.mako.rs b/components/style/properties/longhand/font.mako.rs
index c29ed52..761dd95 100644
--- a/components/style/properties/longhand/font.mako.rs
+++ b/components/style/properties/longhand/font.mako.rs
@@ -1754,7 +1754,7 @@ ${helpers.single_keyword_system("font-variant-position",
                                 spec="https://drafts.csswg.org/css-fonts/#propdef-font-variant-position",
                                 animation_value_type="none")}
 
-<%helpers:longhand name="font-feature-settings" products="none" animation_value_type="none" extra_prefixes="moz"
+<%helpers:longhand name="font-feature-settings" products="gecko" animation_value_type="none" extra_prefixes="moz"
                    spec="https://drafts.csswg.org/css-fonts/#propdef-font-feature-settings">
     use std::fmt;
     use style_traits::ToCss;

```

File: components/style/properties/gecko.mako.rs (hunk 4 above)
Comment (WholeGrainGoats) [inline components/style/properties/gecko.mako.rs]: Really appreciate the help! I'll remember to compile with build-geckolib for next time.

**Comment by Manishearth** [comment]
try at https://treeherder.mozilla.org/#/jobs?repo=try&revision=941fb48f3eb3355eee4683f584d9afb49de079cd

**Pull request merged**